src/controller/tests/data_model/DataModelFixtures.cpp

```diff
@@ -464,5 +485,114 @@ Protocols::InteractionModel::Status ServerClusterCommandExists(const ConcreteCom
     return Status::Success;
 }
 
+CustomDataModel & CustomDataModel::Instance()
+{
+    static CustomDataModel model;
+    return model;
+}
+
+CHIP_ERROR CustomDataModel::ReadAttribute(const ReadAttributeRequest & request, AttributeValueEncoder & encoder)
+{
+    AttributeEncodeState mutableState(&encoder.GetState()); // provide a state copy to start.
+
+#if CHIP_CONFIG_USE_EMBER_DATA_MODEL && CHIP_CONFIG_USE_DATA_MODEL_INTERFACE
+    if ((request.path.mEndpointId < chip::Test::kMockEndpointMin) &&
+        (gReadResponseDirective == ReadResponseDirective::kSendDataResponse) &&
+        (request.path.mClusterId == app::Clusters::UnitTesting::Id) &&
+        (request.path.mAttributeId == app::Clusters::UnitTesting::Attributes::Int16u::Id))
+    {
+        // gInt16uTotalReadCount is a global that keeps changing. Further more, encoding
+        // size differs when moving from 0xFF to 0x100, so encoding sizes in TLV differ.
+        //
+        // This is a HACKISH workaround as it relies that we ember-read before datamodel-read
+        gInt16uTotalReadCount--;
+    }
+#endif // CHIP_CONFIG_USE_EMBER_DATA_MODEL && CHIP_CONFIG_USE_DATA_MODEL_INTERFACE
+
+    CHIP_ERROR err = ReadSingleClusterData(request.subjectDescriptor.value_or(Access::SubjectDescriptor()),
+                                           request.readFlags.Has(ReadFlags::kFabricFiltered), request.path,
+                                           TestOnlyAttributeValueEncoderAccessor(encoder).Builder(), &mutableState);
+
+    // state must survive CHIP_ERRORs as it is used for chunking
+    TestOnlyAttributeValueEncoderAccessor(encoder).SetState(mutableState);
+
+    return err;
+}
+
+CHIP_ERROR CustomDataModel::WriteAttribute(const WriteAttributeRequest & request, AttributeValueDecoder & decoder)
+{
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR CustomDataModel::Invoke(const InvokeRequest & request, chip::TLV::TLVReader & input_arguments, CommandHandler * handler)
+{
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+EndpointId CustomDataModel::FirstEndpoint()
```

Comment:
all this seems really similar to the test interaction-model-api stuff above - would you mind adding some comments about how this is intended to be used vs. the other one?